### DISCUSSION TITLE
preset-icons fill

### DISCUSSION THREAD

**Discussion by tsvayer** [unocss/unocss#1381] [Q&A]
Is there a way to fill icons with color or current color?

`<div class="i-tabler-star fill-red"/>`

**Comment by userquin** [comment]
Use text-red or some variant, text-red-500

**Reply by tsvayer** [reply]
Using text-red will set the stroke but not fill the icon.

**Reply by userquin** [reply]
Try c-red to apply background color: https://uno.antfu.me/?s=colors

**Reply by tsvayer** [reply]
I have checked the implementation of preset-icons and it turns out that applying classes to the element with icon class does not affect the actual svg serialized in data uri. For instance, Tabler icon set has fill='none' and I don't see any way to manipulate it without changing the implementation of preset-icons. There is a concept of transform which could be used to manipulate the svg, but it will only run for the custom loaded sets. Any ideas?

**Reply by userquin** [reply]
It depends on svg content, you can add a transform to desired svg filtering by collection and icon https://github.com/unocss/unocss/tree/main/packages/preset-icons#global-custom-icon-transformation

**Reply by userquin** [reply]
Try this another one, I forgot the transform only apply to custom collections: https://github.com/unocss/unocss/tree/main/packages/preset-icons#global-custom-icon-transformation

**Reply by userquin** [reply]
If still not working, maybe we can add also support to include transform on iconify collections

**Reply by tsvayer** [reply]
@userquin global transform will work, but the question is how to use both filled and non-filled icons in the same project. I imaging something like:

`<div class="i-tabler-star" />` for default non-filled version
`<div class="i-tabler-star?fill" />` for solid/filled version

Does it make any sense?

**Reply by tsvayer** [reply]
Here is one possible implementation:
https://github.com/tsvayer/unocss/commit/45728c83bf3c9ce1604836618928bc322d745620

**Comment by antfu** [comment]
I'd consider this is out-of-scope. Preset-icons only consume the icons data but know nothing about how to interpret them. I'd suggest using an icon set the provide both line and solid versions instead of hacking to the svg.

**Reply by tsvayer** [reply]
I will agree that thinking of the preset-icons it may look like out of scope, but adding support for this special case seems like a lot of benefit.

**Reply by chu121su12** [reply]
It feels too risky [altering the tags](https://uno.antfu.me/play/?html=DwEwlgbgBAxgNgQwM5ILwCIBmcCmAPKACwFpMBXOOKAd1IqrABccBbJYmHAO2YCcoAVmSSMwmAJ4dufdAD4AUFCihIsRCgyNeCLkkwB7XiyhIYCXMQDsABmtzFS5eGjxkadAH0kEAOakwlBwwAFxgxIwIAEa4vMQiCLz2wAD0zrIpaRmQskA&config=JYWwDg9gTgLgBAbwFBzgEwKYDNgDsMDCEuOA5gDQpxhQYDOGMAgjDFMAEYCuMwWAnpVQ16jAJIBjYnSHVaDGAFVcESgF84WKBBBwA5FxUS6dPUiQYAHpFjpsAQy4AbeJhz4iJYKQAUyVFDO9ABccADaVKhhehJcdDA6ALSBThh65IhwUk7QoXq0aHpwagC6smVUdAAW0DCxMHSh-qj6sfFJ1bX1enkwVjCJTqRwfZYD0Pa4pBhwNQBuGFDBowN99k5mqGqyIgqN4ZFyokoqPgCUssLyjCxsnDx8-OeXRwqS0n6HqHQS6xihAEYAHQAJheqAkaFweSqrDAjQA9Aj6CAgdUEZsWmoLlQKqg5vZ2JMGqEIi1EF84CB7HUqj5qbTFmcKeTyVJcPEqTSJFU4ABeLmMqBAhk8nwIgB6AH06HNSIkcE4nIkJBJggizpTUHw4PTuVVmc1WS1aDAuFBcCzjazRVVFqFbYs0U5gBIMHraWEAAwlIGpKYwA3g61VSZoVKhHx4MA8DL4MbMvkAPit1tZ7M5EB4-Lg8ZgUdwMZgZwA3FrrVmYECMLg7vRy2mgYq%2BlAfD4wgBrEqJlMd-l8gV6RKJQyJV3EPSatONrDQACi9jFPgwPdT0%2BNGYgqT9EF8CAw2LL67TGDCAJKOdP56BtDATkX7oRAB0OIqnHz23oAEQlAD8ZzCJ8AGUSgAal-J8AQRCg4C-N8%2BRic1aFrIgcigb8p2PVAAEhsQbVlTXNS1KyPY9tnwtRy0o41yPJWi4AqPCgA&options=N4IgzgLgTglgxhEAuaBXApgXyA). It may/may not introduce xss so proceeed at your own risk.

**Reply by tsvayer** [reply]
That looks great! I was not aware of the concept of variants with all the transformation possibilities. Thank you @chu121su12
